rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+which makes the primitive dependent on the tensors's shapes. Primitive creation is
+an expensive operation, this is why oneDNN has a primitive cache to reduce the
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
+
+Cons:
+* Primitive and primitive descriptor creation overhead will stay
+
+#### Key and Value
+
+The cache implementation will reside in the `common` part of the library. In order
+to abstract away `key` and `value` entities we will need to introduce the
+following abstractions:
+
+```cpp
+namespace kernel_cache {
+// Key
+struct key_t {
+    key_t(key_impl_t *impl);
+    virtual ~key_t() = default;
```

Comment:
It's convenient, and low maintenance.
If your concern is the overhead due to the indirections then I think it's not something to worry about because the overhead is negligible and won't affect the end results.